MapboxCoreNavigation/RouteProgress.swift

```diff
@@ -64,6 +64,40 @@ open class RouteProgress: NSObject {
     @objc public var distanceTraveled: CLLocationDistance {
         return route.legs.prefix(upTo: legIndex).map { $0.distance }.reduce(0, +) + currentLegProgress.distanceTraveled
     }
+    
+    /**
+     Total distance traveled including the traveled distance on previous routes during the same session.
+     */
+    @objc public var totalDistanceTraveled: CLLocationDistance {
+        return previousRouteProgresses().map { $0.distanceTraveled }.reduce(0, +) + distanceTraveled
+    }
+    
+    /**
+     Total distance since the navigation session started, excluding distances when being rerouted.
+     */
+    @objc public var totalDistance: CLLocationDistance {
+        return previousRouteProgresses().map { $0.distanceTraveled }.reduce(0, +) + route.distance
+    }
+    
+    /**
+     Number between 0 and 1 representing how far along the `Route`, the user has traveled, including half completed routes.
+     */
+    @objc public var totalFractionTraveled: Double {
+        return totalDistanceTraveled / totalDistance
+    }
+    
+    // Traverses all previous route progresses, excluding `self`.
+    func previousRouteProgresses() -> [RouteProgress] {
+        var current = self
+        var progresses = [RouteProgress]()
+        
+        while let prev = current.previousRouteProgress {
+            progresses.append(prev)
+            current = prev
```

Comment:
The sheer amount of functional programming in this file makes this non-recursive implementation stick out, but building up this array iteratively is a perfectly valid choice since it’s so predictable. 🙂